src/cinder/FileWatcher.cpp

```diff
@@ -0,0 +1,412 @@
+// FileWatcher
+// ----------------------------------------------------------------------------------------------------
+
+// static
+const FileWatcherRef& FileWatcher::instance()
+{
+	static FileWatcherRef sInstance;
+	if( ! sInstance )
+		sInstance = create();
+
+	return sInstance;
+}
+
+// static
+FileWatcherRef FileWatcher::create()
+{
+	return FileWatcherRef( new FileWatcher );
+}
+
+FileWatcher::FileWatcher()
+{
+}
+
+FileWatcher::~FileWatcher()
+{
+	stopWatchPolling();
+}
+
+void FileWatcher::setWatchingEnabled( bool enable )
+{
+	if( mWatchingEnabled == enable )
+		return;
+
+	mWatchingEnabled = enable;
+	if( enable )
+		instance()->configureWatchPolling();
+	else
+		instance()->stopWatchPolling();
+}
+
+void FileWatcher::setConnectToAppUpdateEnabled( bool enable )
+{
+	if( mConnectToAppUpdateEnabled == enable )
+		return;
+
+	mConnectToAppUpdateEnabled = enable;
+
+	if( ! enable && mConnectionAppUpdate.isConnected() )
+		mConnectionAppUpdate.disconnect();
+}
+
+signals::Connection FileWatcher::watch( const fs::path &filePath, const function<void ( const WatchEvent& )> &callback )
+{ 
+	vector<fs::path> filePaths = { filePath };
+	return watch( filePaths, Options(), callback );
+}
+
+signals::Connection FileWatcher::watch( const fs::path &filePath, const Options &options, const function<void( const WatchEvent& )> &callback )
+{
+	vector<fs::path> filePaths = { filePath };
+	return watch( filePaths, options, callback );
+}
+
+signals::Connection FileWatcher::watch( const vector<fs::path> &filePaths, const function<void ( const WatchEvent& )> &callback )
+{ 
+	return watch( filePaths, Options(), callback );
+}
+
+signals::Connection FileWatcher::watch( const vector<fs::path> &filePaths, const Options &options, const function<void ( const WatchEvent& )> &callback )
+{
+	auto watch = new Watch( filePaths, options.mCallOnWatch );
+	auto conn = watch->connect( callback );
+
+	lock_guard<recursive_mutex> lock( mMutex );
+
+	mWatchList.emplace_back( watch );
+
+	if( options.mCallOnWatch )
+		watch->emitCallback();
+
+	configureWatchPolling();
+
+	return conn;
+}
+
+void FileWatcher::unwatch( const fs::path &filePath )
+{
+	auto fullPath = findFullFilePath( filePath );
+
+	for( auto &watch : mWatchList ) {
+		watch->unwatch( fullPath );
+	}
+}
+
+void FileWatcher::unwatch( const vector<fs::path> &filePaths )
+{
+	for( const auto &filePath : filePaths ) {
+		unwatch( filePath );
+	}
+}
+
+void FileWatcher::enable( const fs::path &filePath )
+{
+	auto fullPath = findFullFilePath( filePath );
+
+	for( auto &watch : mWatchList ) {
+		watch->setEnabled( true, fullPath );
+	}
+}
+
+void FileWatcher::disable( const fs::path &filePath )
+{
+	auto fullPath = findFullFilePath( filePath );
+
+	for( auto &watch : mWatchList ) {
+		watch->setEnabled( false, fullPath );
+	}
+}
+
+void FileWatcher::configureWatchPolling()
+{
+	if( mConnectToAppUpdateEnabled && ! mConnectionAppUpdate.isConnected() && app::App::get() )
```

Comment:
There are app less situations, such as in the unit tests, or if you made a standalone program that did some sort of file post processing or something.